pkg/util/selectors/selectors.go

```diff
@@ -0,0 +1,182 @@
+		}
+	case EqualsOperator, DoubleEqualsOperator, NotEqualsOperator:
+		if len(vals) != 1 {
+			return nil, fmt.Errorf("exact-match compatibility requires one single value")
+		}
+	case ExistsOperator, DoesNotExistOperator:
+		if len(vals) != 0 {
+			return nil, fmt.Errorf("values set must be empty for exists and does not exist")
+		}
+	default:
+		return nil, fmt.Errorf("operator '%v' is not recognized", op)
+	}
+	return &Requirement{key: key, operator: op, strValues: vals}, nil
+}
+
+// Matches returns true if the Requirement matches the input KeyValues.
+// There is a match in the following cases:
+// (1) The operator is Exists and KeyValues has the Requirement's key.
+// (2) The operator is In, KeyValues has the Requirement's key and KeyValues'
+//     value for that key is in Requirement's value set.
+// (3) The operator is NotIn, Labels has the Requirement's key and
+//     KeyValues' value for that key is not in Requirement's value set.
+// (4) The operator is DoesNotExist or NotIn and KeyValues does not have the
+//     Requirement's key.
+func (r *Requirement) Matches(kvs KeyValues) bool {
+	switch r.operator {
+	case InOperator, EqualsOperator, DoubleEqualsOperator:
+		if !kvs.Has(r.key) {
+			return false
+		}
+		return r.strValues.Has(kvs.Get(r.key))
+	case NotInOperator, NotEqualsOperator:
+		if !kvs.Has(r.key) {
+			return true
+		}
+		return !r.strValues.Has(kvs.Get(r.key))
+	case ExistsOperator:
+		return kvs.Has(r.key)
+	case DoesNotExistOperator:
+		return !kvs.Has(r.key)
+	default:
+		return false
+	}
+}
+
+func (r *Requirement) Key() string {
+	return r.key
+}
+func (r *Requirement) Operator() Operator {
+	return r.operator
+}
+func (r *Requirement) Values() sets.String {
+	ret := sets.String{}
+	for k := range r.strValues {
+		ret.Insert(k)
+	}
+	return ret
+}
+
+// String returns a human-readable string that represents this
+// Requirement. If called on an invalid Requirement, an error is
+// returned. See NewRequirement for creating a valid Requirement.
+func (r *Requirement) String() string {
+	var buffer bytes.Buffer
+	if r.operator == DoesNotExistOperator {
+		buffer.WriteString("!")
+	}
+	buffer.WriteString(r.key)
+
+	switch r.operator {
+	case EqualsOperator:
+		buffer.WriteString("=")
+	case DoubleEqualsOperator:
+		buffer.WriteString("==")
+	case NotEqualsOperator:
+		buffer.WriteString("!=")
+	case InOperator:
+		buffer.WriteString(" in ")
+	case NotInOperator:
+		buffer.WriteString(" notin ")
+	case ExistsOperator, DoesNotExistOperator:
+		return buffer.String()
+	}
+
+	switch r.operator {
+	case InOperator, NotInOperator:
+		buffer.WriteString("(")
+	}
+	if len(r.strValues) == 1 {
+		buffer.WriteString(r.strValues.List()[0])
+	} else { // only > 1 since == 0 prohibited by NewRequirement
+		buffer.WriteString(strings.Join(r.strValues.List(), ","))
+	}
+
+	switch r.operator {
+	case InOperator, NotInOperator:
+		buffer.WriteString(")")
+	}
+	return buffer.String()
+}
```

Comment:
I would move to  `pkg/util/selectors` package all the `Requirement` constructor tests that are actually in `pkg/utils/labels` https://github.com/kubernetes/kubernetes/blob/master/pkg/labels/selector_test.go#L287 except the tests with a different validation like for example this: https://github.com/kubernetes/kubernetes/blob/master/pkg/labels/selector_test.go#L302